### PR TITLE
Added support for static plugins

### PR DESCRIPTION
**Pull request by EvanBacon**
TL;DR: [Docs](https://github.com/expo/expo-cli/blob/@evanbacon/config/static-plugins/packages/config-plugins/README.md)

RFC: Maybe it makes sense to support static plugins via JSON as part of the versioning migration. This system implements a plugins array which can point to files that export a `ConfigPlugin` function. Paths can also point to node modules that have a `./app.plugin.js` file in them. 

A plugin can be defined as a string pointing to the module resolution, or an array `[ 'module-path', {} ]`, the props will be passed to the config plugin function: `withPlugin(config, <props>)`.

```js
{
  name: 'my-app',
  plugins: [
    // resolves ./my-custom-plugin.js
    "./my-custom-plugin.js", 
    ["./packages/my-package/expo-plugin.js", { name: 'custom json name 1' }],
    [
      // resolves any node module installed as `react-native-bacon` to `react-native-bacon/app.plugin.js`
      "react-native-bacon", 
       { name: 'custom json name 2' }
    ],
    "react-native-bacon",
  ],
}
```

### CODE REVIEW

**Comment by brentvatne** [comment]
this looks good, and it's great that you were able to implement it using the plugin system itself!

File: packages/config-plugins/README.md [hunk 1]
```diff
@@ -51,7 +51,7 @@ export default withMySDK(config, { apiKey: 'X-XXX-XXX' });
 
 #### Chaining Plugins
 
-Once you add a few plugins, your `app.config.js` code can become difficult to read and manipulate. To combat this, `@expo/config-plugins` provides a `withPlugins` function which can be used to chain plugins together and execute them in order.
+Once you add a few plugins, your `app.plugin.js` code can become difficult to read and manipulate. To combat this, `@expo/config-plugins` provides a `withPlugins` function which can be used to chain plugins together and execute them in order.
```
Comment (EvanBacon) [inline packages/config-plugins/README.md:54]:
```suggestion
Once you add a few plugins, your `app.config.js` code can become difficult to read and manipulate. To combat this, `@expo/config-plugins` provides a `withPlugins` function which can be used to chain plugins together and execute them in order.
```
```suggestion
Once you add a few plugins, your `app.plugin.js` code can become difficult to read and manipulate. To combat this, `@expo/config-plugins` provides a `withPlugins` function which can be used to chain plugins together and execute them in order.
```

Comment (EvanBacon) [inline packages/config-plugins/README.md]:
```suggestion
Once you add a few plugins, your `app.config.js` code can become difficult to read and manipulate. To combat this, `@expo/config-plugins` provides a `withPlugins` function which can be used to chain plugins together and execute them in order.
```

File: packages/config-plugins/README.md [hunk 2]
```diff
@@ -164,3 +164,148 @@ export const withIcons: ConfigPlugin = config => {
...
+```js
+{
+  plugins: [
+    // Long form with properties
+    {
+      resolve: 'my-plugin',
+      props: {
+        /* Passed as the second parameter to the config plugin */
+      },
+    },
```
Comment (EvanBacon) [inline packages/config-plugins/README.md:193]: An alternative API that could be more inline with config plugins is:
```js
{
  plugins: [
    ['my-plugin', { /* props */}],
    './my-other-plugin',
  ]
}
```

This could be extended to work like `withPlugins`

```js

const withSomething = (config, props) => config;

{
  plugins: [
    [withSomething, {}],
    withSomething,
    'expo-splash-screen'
  ]
}
```

File: packages/config-plugins/README.md [hunk 3]
```diff
@@ -164,3 +164,172 @@ export const withIcons: ConfigPlugin = config => {
...
+  plugins: [
+    // Long form with properties
+    ['my-plugin', { /* Passed as the second parameter to the config plugin */ }]
+    // Short hand -- gets normalized to ['my-plugin', undefined]
+    'my-plugin',
+  ];
+}
+```
+
+> 💡 Everything in the Expo config must be able to be converted to JSON (with the exception of the `mods` field). So no async functions!
```
Comment (brentvatne) [inline packages/config-plugins/README.md:195]:
```suggestion
> 💡 Everything in the Expo config must be able to be converted to JSON (with the exception of the `mods` field). So no async functions outside of `mods` in your config plugins!
```

File: packages/config-plugins/README.md [hunk 4]
```diff
@@ -164,3 +164,172 @@ export const withIcons: ConfigPlugin = config => {
...
+    ├── package.json ➡️ The `main` file points to `build/index.js`
+    ╰── build/index.js ➡️  ✅ Node resolves to this module.
+```
+
+Sometimes you want your package to export React components and also support a plugin, to support this, multiple entry points are used. If a `app.plugin.js` file is present in the Node module's root folder, it'll be used instead of the package's main file.
+
+```
+╭── app.config.js ➡️ Expo Config
+╰── node_modules/expo-splash-screen/ ➡️ Module installed from NPM (works with Yarn workspaces as well).
+    ├── package.json ➡️ The `main` file will be used if `index.expo-plugin.js` doesn't exist.
```
Comment (brentvatne) [inline packages/config-plugins/README.md:228]:
```suggestion
    ├── package.json ➡️ The `main` file will be used if `app.plugin.js` doesn't exist.
```

File: packages/config-plugins/README.md [hunk 5]
```diff
@@ -164,3 +164,172 @@ export const withIcons: ConfigPlugin = config => {
...
+`'expo-splash-screen'`
+
+```
+╭── app.config.js ➡️ Expo Config
+╰── node_modules/expo-splash-screen/ ➡️ Module installed from NPM (works with Yarn workspaces as well).
+    ├── package.json ➡️ The `main` file points to `build/index.js`
+    ╰── build/index.js ➡️  ✅ Node resolves to this module.
+```
+
+Sometimes you want your package to export React components and also support a plugin, to support this, multiple entry points are used. If a `app.plugin.js` file is present in the Node module's root folder, it'll be used instead of the package's main file.
```
Comment (brentvatne) [inline packages/config-plugins/README.md:223]: maybe we should move this above the `build/index.js` example, because it takes precedence

File: packages/config-plugins/README.md [hunk 6]
```diff
@@ -164,3 +164,172 @@ export const withIcons: ConfigPlugin = config => {
...
+
+```
+╭── app.config.js ➡️ Expo Config
+╰── node_modules/expo-splash-screen/ ➡️ Module installed from NPM (works with Yarn workspaces as well).
+    ├── package.json ➡️ The `main` file will be used if `index.expo-plugin.js` doesn't exist.
+    ├── app.plugin.js ➡️ ✅ `module.exports = (config) => config` -- must export a function.
+    ╰── build/index.js ➡️ ❌ Ignored because `index.expo-plugin.js` exists. This could be used with `expo-splash-screen/build/index.js`
```
Comment (brentvatne) [inline packages/config-plugins/README.md:230]:
```suggestion
    ╰── build/index.js ➡️ ❌ Ignored because `app.plugin.js` exists. This could be used with `expo-splash-screen/build/index.js`
```

File: packages/config-plugins/README.md [hunk 7]
```diff
@@ -164,3 +164,172 @@ export const withIcons: ConfigPlugin = config => {
...
+```
+
+#### Project folder
+
+`'./my-config-plugin'`
+
+This is different to how Node module's work because `app.plugin.js` won't be resolved by default in a folder. You'll have to manually specify the file to use it.
```
Comment (brentvatne) [inline packages/config-plugins/README.md:237]:
```suggestion
This is different to how Node modules work because `app.plugin.js` won't be resolved by default in a directory. You'll have to manually specify `./my-config-plugin/app.plugin.js` to use it, otherwise `index.js` in the directory will be used.
```

File: packages/config-plugins/README.md [hunk 8]
```diff
@@ -164,3 +164,172 @@ export const withIcons: ConfigPlugin = config => {
...
+
+```
+╭── app.config.js ➡️ Expo Config
+╰── my-config-plugin/ ➡️ Folder containing plugin code
+    ╰── index.js ➡️ ✅ By default, node resolves a folder's index.js file as the main file.
```
Comment (brentvatne) [inline packages/config-plugins/README.md:242]:
```suggestion
    ╰── index.js ➡️ ✅ By default, Node resolves a folder's index.js file as the main file.
```

for consistency with casing above

File: packages/config-plugins/README.md [hunk 9]
```diff
@@ -164,3 +164,172 @@ export const withIcons: ConfigPlugin = config => {
...
+#### Module internals
+
+If a file inside a Node module is specified, then the module's root `app.plugin.js` resolution will be skipped. This is referred to as "reaching inside a package" and is bad form. We support this to make testing, and plugin authoring easier.
+
+- `'expo-splash-screen/build/index.js'`
+- `'expo-splash-screen/build'`
+
+```
+╭── app.config.js ➡️ Expo Config
+╰── node_modules/expo-splash-screen/ ➡️ Module installed from NPM (works with Yarn workspaces as well).
```
Comment (brentvatne) [inline packages/config-plugins/README.md:254]:
```suggestion
╰── node_modules/expo-splash-screen/ ➡️ Module installed from npm (works with Yarn workspaces as well).
```

npm spells their company name in all lower case

File: packages/config-plugins/README.md [hunk 10]
```diff
@@ -164,3 +164,172 @@ export const withIcons: ConfigPlugin = config => {
...
+
+```
+╭── app.config.js ➡️ Expo Config
+╰── my-config-plugin/ ➡️ Folder containing plugin code
+    ╰── index.js ➡️ ✅ By default, node resolves a folder's index.js file as the main file.
+```
+
+#### Module internals
+
+If a file inside a Node module is specified, then the module's root `app.plugin.js` resolution will be skipped. This is referred to as "reaching inside a package" and is bad form. We support this to make testing, and plugin authoring easier.
```
Comment (brentvatne) [inline packages/config-plugins/README.md:247]:
```suggestion
If a file inside a Node module is specified, then the module's root `app.plugin.js` resolution will be skipped. This is referred to as "reaching inside a package" and is bad form. We support this to make testing, and plugin authoring easier, but we don't expect library authors to expose their plugins like this as a public API.
```

File: packages/config-plugins/README.md [hunk 11]
```diff
@@ -164,3 +164,172 @@ export const withIcons: ConfigPlugin = config => {
...
+
+- `'expo-splash-screen/build/index.js'`
+- `'expo-splash-screen/build'`
+
+```
+╭── app.config.js ➡️ Expo Config
+╰── node_modules/expo-splash-screen/ ➡️ Module installed from NPM (works with Yarn workspaces as well).
+    ├── package.json ➡️ The `main` file will be used if `index.expo-plugin.js` doesn't exist.
```
Comment (brentvatne) [inline packages/config-plugins/README.md:255]:
```suggestion
    ├── package.json ➡️ The `main` file will be used if `app.plugin.js` doesn't exist.
```

File: packages/config-plugins/README.md [hunk 12]
```diff
@@ -164,3 +164,172 @@ export const withIcons: ConfigPlugin = config => {
...
+```json
+{
+  "plugins": [["withCustom", {}], "withCustom"]
+}
+```
+
+### Why app.plugin.js for plugins
+
+Config resolution searches for a `app.plugin.js` first when a Node module name is provided.
+This is because node environments are often different to iOS, Android, or web JS environments and therefore require different transpilation presets (ex: `module.exports` instead of `import/export`).
```
Comment (brentvatne) [inline packages/config-plugins/README.md:292]:
```suggestion
This is because Node environments are often different to iOS, Android, or web JS environments and therefore require different transpilation presets (ex: `module.exports` instead of `import/export`).
```

File: packages/config-plugins/README.md [hunk 13]
```diff
@@ -164,3 +164,172 @@ export const withIcons: ConfigPlugin = config => {
...
+
+Because of this reasoning, the root of a Node module is searched instead of right next to the `index.js`. Imagine you had a TypeScript Node module where the transpiled main file was located at `build/index.js`, if Expo config plugin resolution searched for `build/app.plugin.js` you'd lose the ability to transpile the file differently.
+
+### Creating static plugins
+
+- The root file can be any JS file or the root `app.plugin.js` in a Node module.
+- The file should export a `ConfigPlugin` function.
+- Plugins should be transpiled for Node environments ahead of time!
```
Comment (brentvatne) [inline packages/config-plugins/README.md:300]: maybe we should recommend supporting all LTS releases and the current release of node?

File: packages/config-plugins/README.md [hunk 14]
```diff
@@ -164,3 +164,172 @@ export const withIcons: ConfigPlugin = config => {
...
+### Why app.plugin.js for plugins
+
+Config resolution searches for a `app.plugin.js` first when a Node module name is provided.
+This is because node environments are often different to iOS, Android, or web JS environments and therefore require different transpilation presets (ex: `module.exports` instead of `import/export`).
+
+Because of this reasoning, the root of a Node module is searched instead of right next to the `index.js`. Imagine you had a TypeScript Node module where the transpiled main file was located at `build/index.js`, if Expo config plugin resolution searched for `build/app.plugin.js` you'd lose the ability to transpile the file differently.
+
+### Creating static plugins
+
+- The root file can be any JS file or the root `app.plugin.js` in a Node module.
```
Comment (brentvatne) [inline packages/config-plugins/README.md:298]: not sure what the right wording here is but i think this is the first time mentioning "root file" so we might want to reword this a bit. maybe like "The entry point to your plugin can be ..."

**Review by brentvatne — COMMENTED** [review]
readme comments

File: packages/config/src/plugins/modulePluginResolver.ts [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+function resolvePluginForModule(projectRoot: string, modulePath: string): string {
+  const resolved = resolveFrom(projectRoot, modulePath);
+  assert(
+    resolved,
+    `Failed to resolve plugin for module "${modulePath}" relative to "${projectRoot}"`
+  );
+  // If the modulePath is something like `@bacon/package/index.js` or `expo-foo/build/app`
+  // then skip resolving the module `index.expo-plugin.js`
```
Comment (brentvatne) [inline packages/config/src/plugins/modulePluginResolver.ts:28]:
```suggestion
  // then skip resolving the module `app.plugin.js`
```

File: packages/config/src/Serialize.ts [hunk 16]
```diff
@@ -38,3 +38,42 @@ export function serializeSkippingMods(val: any): any {
   }
   return serializeAndEvaluate(val);
 }
+
+function serializeAndEvaluatePlugin(val: any): any {
+  if (['undefined', 'string', 'boolean', 'number', 'bigint'].includes(typeof val)) {
+    return val;
+  } else if (typeof val === 'function') {
+    // TODO: Attempt to replace the function with it's name
```
Comment (brentvatne) [inline packages/config/src/Serialize.ts:46]:
```suggestion
```

**Review by brentvatne — APPROVED** [review]
looks good. as discussed, should rework readme to move the 'how to use' before 'how it works' part. i like this api, feels like it's getting to a really nice spot

File: packages/config-plugins/README.md [hunk 17]
```diff
@@ -164,3 +164,172 @@ export const withIcons: ConfigPlugin = config => {
...
+Config resolution searches for a `app.plugin.js` first when a Node module name is provided.
+This is because Node environments are often different to iOS, Android, or web JS environments and therefore require different transpilation presets (ex: `module.exports` instead of `import/export`).
+
+Because of this reasoning, the root of a Node module is searched instead of right next to the `index.js`. Imagine you had a TypeScript Node module where the transpiled main file was located at `build/index.js`, if Expo config plugin resolution searched for `build/app.plugin.js` you'd lose the ability to transpile the file differently.
+
+### Creating static plugins
+
+- The root file can be any JS file or the root `app.plugin.js` in a Node module.
+- The file should export a `ConfigPlugin` function.
+- Plugins should be transpiled for Node environments ahead of time!
```
Comment (EvanBacon) [inline packages/config-plugins/README.md:300]:
```suggestion
- Plugins should be transpiled for Node environments ahead of time! They should support the versions of Node that [Expo supports](https://docs.expo.io/get-started/installation/#requirements) (LTS).
```

File: packages/config-plugins/README.md [hunk 18]
```diff
@@ -2,17 +2,85 @@
 
 The Expo config is a powerful tool for generating native app code from a unified JavaScript interface. Most basic functionality can be controlled by using the the [static Expo config](https://docs.expo.io/versions/latest/config/app/), but some features require manipulation of the native project files. To support complex behavior we've created config plugins, and mods (short for modifiers).
 
-> Here is a [colorful chart](https://whimsical.com/UjytoYXT2RN43LywvWExfK) for visual learners.
+> 💡 **Hands-on Learners**: Use [this sandbox][sandbox] to play with the core functionality of Expo config. For more complex tests, use a local Expo project, with `expo eject --no-install` to apply changes.
```
Comment (brentvatne) [inline packages/config-plugins/README.md:5]:
```suggestion
> 💡 **Hands-on Learners**: Use [this sandbox][sandbox] to play with the core functionality of Expo config plugins. For more complex tests, use a local Expo project, with `expo eject --no-install` to apply changes.
```

File: packages/config-plugins/README.md [hunk 19]
```diff
@@ -2,17 +2,85 @@
...
+  - [Module internals](#module-internals)
+  - [Raw functions](#raw-functions)
+- [Why app.plugin.js for plugins](#why-apppluginjs-for-plugins)
+
+**Quick facts**
+
+- Plugins are functions that can change values on your Expo config.
+- Plugins are mostly meant to be used with [`expo eject`][cli-eject] or `eas build` commands.
+- We recommend you use plugins with `app.config.json` or `app.config.js` instead of `app.json` (no top-level `expo` object is required).
+- `mods` are async functions that modify native files.
```
Comment (brentvatne) [inline packages/config-plugins/README.md:32]:
```suggestion
- `mods` are async functions that modify native project files, such as source code or configuration (plist, xml) files.
```

File: packages/config-plugins/README.md [hunk 20]
```diff
@@ -2,17 +2,85 @@
...
+- Changes performed with `mods` will require a native rebuild.
```
Comment (brentvatne) [inline packages/config-plugins/README.md:33]:
```suggestion
- Changes performed with `mods` will require rebuilding the affected native projects.
```

File: packages/config-plugins/README.md [hunk 21]
```diff
@@ -49,7 +117,56 @@ const config = {
 export default withMySDK(config, { apiKey: 'X-XXX-XXX' });
 ```
 
-#### Chaining Plugins
+### Importing plugins
+
+You may want to create a plugin in a different file, here's how:
+
+- The root file can be any JS file or a file named `app.plugin.js` in the [root of a Node module](#root-app.plugin.js).
+- The file should export a `ConfigPlugin` function.
```
Comment (brentvatne) [inline packages/config-plugins/README.md:125]:
```suggestion
- The file should export a function that satisfies the `ConfigPlugin` type.
```

Comment (brentvatne) [reply packages/config-plugins/README.md:125]: should we add the type definition to this readme or link to it?

File: packages/config-plugins/README.md [hunk 22]
```diff
@@ -68,39 +185,65 @@ import { withPlugins } from '@expo/config-plugins';
...
+    },
+    android: {
+      /* Android mods... */
+    },
+  },
+};
 ```
 
-### How it works
+## How mods works
```
Comment (brentvatne) [inline packages/config-plugins/README.md:233]:
```suggestion
## How mods work
```

**Review by brentvatne — APPROVED** [review]
